### PR TITLE
drivers: nrfx: fix USB in endpoint data race

### LINKED ISSUE #71299: USB core has a data race during write on many platforms
The function `hid_int_ep_write()` has a data race where the TX data may change underneath the USB driver. For example, `hid_int_ep_write()` calls `usb_write()` directly passing the user's data buffer. Many USB drivers do not copy the buffer before using it e.g. `nrfx` and `stm32`, while e.g. `rpi_pico`, `native_posix`, and `kinetis` do. This means that if the caller modifies the passed data buffer before the TX is complete the host may observe the new contents.

The problem was identified on the `nrfx` platform (a Kinesis Adv360 keyboard running ZMK), and was observed to very reliably cause keys to be dropped. On wireshark two TX messages were recorded with each keypress (corresponding to key press and key release), but both messages contained the same contents (no keys pressed). Only a key press-release combination generated by a macro-like mode of ZMK is fast enough to trigger the bug. My proposed fix to ZMK, the PR https://github.com/zmkfirmware/zmk/pull/2257, simply copies the data into a temporary buffer before the call and immediately fixed the problem as tested on real hardware.

I am confident this is bug in Zephyr because all but one of the samples which call `hid_int_ep_write()`, namely [`usb/hid-mouse`](https://github.com/zephyrproject-rtos/zephyr/blob/b9f3d68863b9f75dcc1dfac7dd7ff269b99cd118/samples/subsys/usb/hid-mouse/src/main.c#L133), [`usb/hid-cdc`](https://github.com/zephyrproject-rtos/zephyr/blob/b9f3d68863b9f75dcc1dfac7dd7ff269b99cd118/samples/subsys/usb/hid-cdc/src/main.c#L659), and [`sensor/fxos8700-hid`](https://github.com/zephyrproject-rtos/zephyr/blob/b9f3d68863b9f75dcc1dfac7dd7ff269b99cd118/samples/sensor/fxos8700-hid/src/main.c#L246) all do not protect the data buffer after calling `hid_int_ep_write()`, e.g. by modifying it immediately after the call, or by leaving it on the stack and returning allowing for corruption. (Note that the semaphore in `sensor/fxos8700-hid` is not protecting the data buffer.) The only sample which does not have this issue ([`usb/hid`](https://github.com/zephyrproject-rtos/zephyr/blob/b9f3d68863b9f75dcc1dfac7dd7ff269b99cd118/samples/subsys/usb/hid/src/main.c#L63)) is because it is too simple and uses a constant buffer.

In order to fix the problem either the HID function `hid_int_ep_write()` or the USB core function `usb_write()` (or the underlying driver `usb_write()` immediately calls) need to copy the data buffer. I am happy to implement a fix in any case but seek confirmation of what the intended API is and thus which approach to take.

Thank you very much in advance for your help.

### PR DESCRIPTION
**Pull request by khoek**
The function `usb_dc_ep_write()` races against its caller because it does not copy the passed data buffer as expected by its contract, and as the other drivers do. Thus the TX data may change from underneath the driver while it is pending.

Alongside the output endpoint RX buffers which already exist, we define an input endpoint TX buffer for each endpoint and copy into TX data into it before transmitting. The new buffer is protected by the same lock which prevents a write being issued while an existing write is in progress.

This bug was discovered on a Kinesis Adv360 keyboard running ZMK, and was observed to very reliably cause keys with a sufficiently high keycode (hence last to be transmitted) to be dropped. With Wireshark two TX messages were recorded on each keypress (corresponding to key press and key release), but both messages contained the same contents (no keys pressed). Only a key press-release combination generated by a macro-like mode of ZMK is fast enough to trigger the bug. My proposed fix to ZMK, the PR https://github.com/zmkfirmware/zmk/pull/2257, simply copies the data into a temporary buffer before the call and immediately fixed the problem.

This commit also fixes the bug now using a vanilla copy of ZMK, and has been tested to work on real hardware when backported to ZMK's Zephyr fork.

Closes #71299.

Signed-off-by: Keeley Hoek <keeley@hoek.io>

### CODE REVIEW

**Review by tmon-nordic — COMMENTED** [review]
Thank you for the excellent PR and issue report. I do agree that copy would absolutely eliminate the race. Unfortunately the significant additional RAM usage will likely break many applications and therefore I am hesistant about this solution.

The API is not really claiming that it will copy the data (please provide the exact place where it says so if I am wrong) but at the same time does not restrain caller to keep the buffer intact (this is what I would have implied).

Would it be possible to change the fix to only affect `hid_int_ep_write()` where you observed the issue? I think other callers are fine with the requirements for buffer to stay intact. Doing it only for HID endpoints would significantly reduce the memory usage.

**Comment by khoek** [comment]
Thank you so much for your amazingly fast review. Modifying `hid_int_ep_write()` instead was exactly my first idea too, but  the fact that all of the USB samples [`usb/hid-mouse`](https://github.com/zephyrproject-rtos/zephyr/blob/b9f3d68863b9f75dcc1dfac7dd7ff269b99cd118/samples/subsys/usb/hid-mouse/src/main.c#L133), [`usb/hid-cdc`](https://github.com/zephyrproject-rtos/zephyr/blob/b9f3d68863b9f75dcc1dfac7dd7ff269b99cd118/samples/subsys/usb/hid-cdc/src/main.c#L659), and [`sensor/fxos8700-hid`](https://github.com/zephyrproject-rtos/zephyr/blob/b9f3d68863b9f75dcc1dfac7dd7ff269b99cd118/samples/sensor/fxos8700-hid/src/main.c#L246) (all do not protect the data buffer after calling `hid_int_ep_write()`) would still be broken steered me away from this path. (Note that the semaphore in `sensor/fxos8700-hid` is not protecting the data buffer.) The only sample which does not have this issue ([`usb/hid`](https://github.com/zephyrproject-rtos/zephyr/blob/b9f3d68863b9f75dcc1dfac7dd7ff269b99cd118/samples/subsys/usb/hid/src/main.c#L63)) is because it is too simple and uses a constant buffer. In addition to these and `usb_dc_ep_write()` in `usb/hid/core.c`, I also found other files such as https://github.com/zephyrproject-rtos/zephyr/blob/731f8d473610f04ec3d0cf53cb10dd373b467f31/subsys/usb/device/class/msc.c#L445-L450 which build a bunch of buffers on the stack and would also be broken.

On the other hand: https://github.com/zephyrproject-rtos/zephyr/blob/731f8d473610f04ec3d0cf53cb10dd373b467f31/subsys/usb/device/usb_transfer.c#L239 appears to me to also expect that the buffer is not touched until the application finds out it is done when the passed `usb_transfer_callback cb` is invoked, given its approach to chunking.

Also, from my attempt at reading all of the other drivers, I believe all others copy the buffer, e.g.
https://github.com/zephyrproject-rtos/zephyr/blob/9c05618c2b6b8e10ab523411067e32846cd8cd8e/drivers/usb/device/usb_dc_rpi_pico.c#L112-L114
or are synchronous, e.g.
https://github.com/zephyrproject-rtos/zephyr/blob/9c05618c2b6b8e10ab523411067e32846cd8cd8e/drivers/usb/device/usb_dc_stm32.c#L846-L860

Of course, if it is technically infeasible to make a change along these lines due to breaking existing users I understand. But there certainly is a problem to solve somewhere. (Another solution with no memory footprint would of course be to make the call synchronous, but I would imagine that could be a significant breaking change as well.)

File: drivers/usb/device/usb_dc_nrfx.c [hunk 1]
```diff
@@ -1689,8 +1701,13 @@ int usb_dc_ep_write(const uint8_t ep, const uint8_t *const data,
 		return 0;
 	}
 
+	if (data_len > EP_IN_BUF_MAX_SZ) {
+		data_len = EP_IN_BUF_MAX_SZ;
+	}
+	memcpy(ep_ctx->in_buf, data, data_len);
```
[the remark this answers is not in the document]
Comment (khoek) [reply drivers/usb/device/usb_dc_nrfx.c:1707]: I see. That sounds like a serious problem.

File: drivers/usb/device/usb_dc_nrfx.c [hunk 2]
```diff
@@ -202,17 +203,26 @@ K_MEM_SLAB_DEFINE(fifo_elem_slab, FIFO_ELEM_SZ,
...
+#define EP_OUT_BUF_MAX_SZ	64UL
+#define ISO_EP_OUT_BUF_MAX_SZ	1024UL
+
+/**
+ * @brief Input endpoint buffers
+ *	Used as buffers for the input endpoints' data transfer
+ *	Max buffers size possible: 1024 Bytes
+ */
+static uint8_t ep_in_bufs[CFG_EPIN_CNT][EP_IN_BUF_MAX_SZ]
+	       __aligned(sizeof(uint32_t));
```
[the remark this answers is not in the document]
Comment (khoek) [reply drivers/usb/device/usb_dc_nrfx.c:216]: Pending your other comment which sounds like a showstopper, could we perhaps allocate the buffers for each EP on first use?

**Comment by tmon-nordic** [comment]
Wouldn't just copying the data in `hid_int_ep_write()` to internal HID instance report buffer be sufficient to fix it?

Note that the USB stack you modify has multiple design issues and patching it is going to be really hard. The drivers are of much varying quality and due to USB DC API not being perfectly clear driver behavior varies (e.g. with regard to copy or no-copy). For that reason there is an effort to write new USB stack (currently marked as experimental; HID is not yet merged to Zephyr main).

**Comment by khoek** [comment]
Yes it would fix my problem, and if you would accept such a patch I will supply it.

File: drivers/usb/device/usb_dc_nrfx.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (tmon-nordic) [reply drivers/usb/device/usb_dc_nrfx.c:216]: That would require use of dynamic memory allocation, which is another can of worms. I'd consider using system heap as even worse.

**Comment by jfischer-no** [comment]
> The function `usb_dc_ep_write()` races against its caller because it does not copy the passed data buffer as expected by its contract, and as the other drivers do. Thus the TX data may change from underneath the driver while it is pending.

It is up to the driver (maybe also depending on the controller and HAL madness) to copy or not, at least the API says that "The supplied usb_ep_callback function will be called when data is transmitted out", which can be interpreted as the caller may not modify the data until e.g. in_ready_cb() is called. Also, modifying hid_int_ep_write() does not make sense, because if the application is designed correctly, it would add an unnecessary buffer.

**Review by jfischer-no — CHANGES_REQUESTED** [review]
See https://github.com/zephyrproject-rtos/zephyr/pull/71306#issuecomment-2047154160

**Comment by tmon-nordic** [comment]
> Also, modifying hid_int_ep_write() does not make sense, because if the application is designed correctly, it would add an unnecessary buffer.

Then we should at least have proper samples in Zephyr. Having bad samples in Zephyr does not really support the "if the application is designed correctly".

**Comment by petejohanson** [comment]
> > Also, modifying hid_int_ep_write() does not make sense, because if the application is designed correctly, it would add an unnecessary buffer.
> 
> Then we should at least have proper samples in Zephyr. Having bad samples in Zephyr does not really support the "if the application is designed correctly".

Fully agreed on this. The samples are where many users (myself included) of Zephyr first look at how to do things with the APIs. They certainly shouldn't be complex to the point of scope creep, but should at a minimum show the correct way to properly use the API for real world use.

**Pull request closed without merging**